x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/DataFrameRowsJoiner.java

```diff
@@ -37,12 +38,12 @@
     private final String analyticsId;
     private final DataFrameDataExtractor dataExtractor;
     private final ResultsPersisterService resultsPersisterService;
+    private Supplier<Boolean> shouldRetryPersistence = () -> true;
```

Comment:
Could this be final and passed in the constructor?

Reply:
Since the retry persistence logic depends on information stored within `AnalyticsResultProcessor` we would either have to make a Builder or a Factory class that we then pass to `AnalyticsResultProcessor` which constructs a new joiner from that passed factory. 

This option just seemed simpler and less complex.

Reply:
Ok, got it. Please also consider the following approach which might be even less complex:

Add `isCancelled` field and `cancel()` method to `DataFrameRowsJoiner` class. In `AnalyticsResultProcessor::cancel()`, you can also cancel the joiner:
```
    public void cancel() {
        isCancelled = true;
        dataFrameRowsJoiner.cancel();
    }
``` 

Then there is no need to pass the `Supplier` as you've already passed the actual `isCancelled` value.